api/trading_data_v2.go

```diff
@@ -193,6 +195,143 @@ func entityMarketDataListToProtoList(list []entities.MarketData) (*v2.MarketData
+	// Wrap context from the request into cancellable. We can close internal chan on error.
+	ctx, cancel := context.WithCancel(srv.Context())
+	defer cancel()
+
+	for _, marketId := range req.MarketIds {
+		_, err := validateMarketSQL(ctx, marketId, t.marketService)
+		if err != nil {
+			return err // validateMarket already returns an API error, no additional wrapping needed
+		}
+	}
+
+	depthChan, ref := t.marketDepthService.ObserveDepth(
+		ctx, t.config.StreamRetries, req.MarketIds)
+
+	if t.log.GetLevel() == logging.DebugLevel {
+		t.log.Debug("Depth subscriber - new rpc stream", logging.Uint64("ref", ref))
+	}
+
+	return observeBatch(ctx, t.log, "MarketDepth", depthChan, ref, func(tr []*vega.MarketDepth) error {
+		return srv.Send(&v2.ObserveMarketsDepthResponse{
+			MarketDepth: tr,
+		})
+	})
+
+}
+
+func (t *tradingDataServiceV2) ObserveMarketsDepthUpdates(req *v2.ObserveMarketsDepthUpdatesRequest, srv v2.TradingDataService_ObserveMarketsDepthUpdatesServer) error {
+	// Wrap context from the request into cancellable. We can close internal chan on error.
+	ctx, cancel := context.WithCancel(srv.Context())
+	defer cancel()
+
+	for _, marketId := range req.MarketIds {
+		_, err := validateMarketSQL(ctx, marketId, t.marketService)
+		if err != nil {
+			return err // validateMarket already returns an API error, no additional wrapping needed
+		}
+	}
+	depthChan, ref := t.marketDepthService.ObserveDepthUpdates(
+		ctx, t.config.StreamRetries, req.MarketIds)
+
+	if t.log.GetLevel() == logging.DebugLevel {
+		t.log.Debug("Depth updates subscriber - new rpc stream", logging.Uint64("ref", ref))
+	}
+
+	return observeBatch(ctx, t.log, "MarketDepthUpdate", depthChan, ref, func(tr []*vega.MarketDepthUpdate) error {
+		return srv.Send(&v2.ObserveMarketsDepthUpdatesResponse{
+			Update: tr,
+		})
+	})
+}
+
+func (t *tradingDataServiceV2) ObserveMarketsData(req *v2.ObserveMarketsDataRequest, srv v2.TradingDataService_ObserveMarketsDataServer) error {
+	// Wrap context from the request into cancellable. We can close internal chan on error.
+	ctx, cancel := context.WithCancel(srv.Context())
+	defer cancel()
+
+	ch, ref := t.marketDataService.ObserveMarketData(ctx, t.config.StreamRetries, req.MarketIds)
+
+	return observeBatch(ctx, t.log, "MarketsData", ch, ref, func(marketData []*entities.MarketData) error {
+		out := make([]*vega.MarketData, 0, len(marketData))
+		for _, v := range marketData {
+			out = append(out, v.ToProto())
+		}
+		return srv.Send(&v2.ObserveMarketsDataResponse{MarketData: out})
+	})
+
+}
+
+// GetLatestMarketData returns the latest market data for a given market
+func (t *tradingDataServiceV2) GetLatestMarketData(ctx context.Context, req *v2.GetLatestMarketDataRequest) (*v2.GetLatestMarketDataResponse, error) {
+	defer metrics.StartAPIRequestAndTimeGRPC("GetLatestMarketData")()
+
+	// validate the market exist
+	if req.MarketId != "" {
```

Comment:
What should we do if string is empty?

Reply:
yeah, that is total shite, just changed it to validate the id, it was erroring anyway if empty string, just a bit further down the stack (market_data.go line 87)